fw/str.c

```diff
@@ -670,6 +688,181 @@ tfw_str_add_duplicate(TfwPool *pool, TfwStr *str)
 	return dup_str;
 }
 
+/**
+ * Append the data and terminator strings as leaves into 2-level tree @array.
+ * If assigned, @terminator will be appended to the last incomplete item.
+ * If @terminator is assigned but @terminator_len is 0, the last item in the
+ * @array is marked as completed, but no terminator is appended
+ * i.e. empty terminator.
+ *
+ * Constracts:
+ * 1. @data != NULL
+ * 2. @terminator == NULL iff @terminator_len == 0
+ * 3. In case of error the caller do not proceed with the HTTP message and
+ *    frees the associated memory pool soon.
+ *
+ * The array is at most 3 levels:
+ * 1. TfwStr for the root
+ * 2. TfwStr's consisting the array (the array items)
+ * 3. string chunks for the arrays items (if the strings are chunked)
+ *
+ * array -> chunks[0] = { nchunks = 2,
+ *                        len = 0,
+ *                        chunks = ["token1", ":"],    <= @terminator was ":"
+ *                        flags = TFW_STR_COMPLETE }
+ *          chunks[1] = { nchunks = 3,
+ *                        len = undefined,
+ *                        chunks = ["tok", "en2", ":"],
+ *                        flags = TFW_STR_COMPLETE }
+ *          chunks[2] = { nchunks = 0,    <= single chunk stored inside parent
+ *                        len = 3,
+ *                        data = "tok",
+ *                        flags = 0 }     <= last uncompleted item
+ *          nchunks = 3
```

Comment:
Please update the comment after the simplification. When the function has signature `tfw_str_array_append_chunk(TfwPool *pool, TfwStr *array, char *data, unsigned long len)` that won't be obvious how it's different from `tfw_str_add_compound()`/`__str_grow_tree()`, so please address this question (why do we need the new function) in the comment.